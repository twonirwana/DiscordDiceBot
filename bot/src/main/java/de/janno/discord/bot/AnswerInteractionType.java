package de.janno.discord.bot;

public enum AnswerInteractionType {
    none,
    hidden,
    reroll
}
