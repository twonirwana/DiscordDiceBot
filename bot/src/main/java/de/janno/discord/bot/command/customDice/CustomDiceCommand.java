package de.janno.discord.bot.command.customDice;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.Lists;
import de.janno.discord.bot.BotMetrics;
import de.janno.discord.bot.command.*;
import de.janno.discord.bot.dice.Dice;
import de.janno.discord.bot.dice.DiceParser;
import de.janno.discord.bot.dice.DiceParserSystem;
import de.janno.discord.bot.dice.DiceSystemAdapter;
import de.janno.discord.bot.persistance.Mapper;
import de.janno.discord.bot.persistance.MessageDataDAO;
import de.janno.discord.bot.persistance.MessageDataDTO;
import de.janno.discord.connector.api.BottomCustomIdUtils;
import de.janno.discord.connector.api.ButtonEventAdaptor;
import de.janno.discord.connector.api.message.ButtonDefinition;
import de.janno.discord.connector.api.message.ComponentRowDefinition;
import de.janno.discord.connector.api.message.EmbedDefinition;
import de.janno.discord.connector.api.message.MessageDefinition;
import de.janno.discord.connector.api.slash.CommandDefinitionOption;
import de.janno.discord.connector.api.slash.CommandInteractionOption;
import de.janno.evaluator.dice.random.NumberSupplier;
import de.janno.evaluator.dice.random.RandomNumberSupplier;
import lombok.NonNull;
import lombok.Value;
import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.Nullable;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

@Slf4j
public class CustomDiceCommand extends AbstractCommand<CustomDiceConfig, StateData> {

    private static final String COMMAND_NAME = "custom_dice";
    private static final List<String> LEGACY_DICE_COMMAND_OPTIONS_IDS = IntStream.range(1, 25).mapToObj(i -> i + "_button").toList();
    private static final String LEGACY_START_ACTION = "legacy_start";
    private static final String BUTTONS_COMMAND_OPTIONS_ID = "buttons";
    private static final String LABEL_DELIMITER = "@";
    private static final String BUTTON_MESSAGE = "Click on a button to roll the dice";
    private static final String CONFIG_TYPE_ID = "CustomDiceConfig";
    private final DiceSystemAdapter diceSystemAdapter;

    public CustomDiceCommand(MessageDataDAO messageDataDAO) {
        this(messageDataDAO, new DiceParser(), new RandomNumberSupplier(), 1000);
    }

    @VisibleForTesting
    public CustomDiceCommand(MessageDataDAO messageDataDAO, Dice dice, NumberSupplier numberSupplier, int maxRolls) {
        super(messageDataDAO);
        this.diceSystemAdapter = new DiceSystemAdapter(numberSupplier, maxRolls, dice);
    }

    @Override
    protected Optional<ConfigAndState<CustomDiceConfig, StateData>> getMessageDataAndUpdateWithButtonValue(long channelId,
                                                                                                           long messageId,
                                                                                                           @NonNull String buttonValue,
                                                                                                           @NonNull String invokingUserName) {
        final Optional<MessageDataDTO> messageDataDTO = messageDataDAO.getDataForMessage(channelId, messageId);
        if (messageDataDTO.isEmpty()) {
            return Optional.empty();
        }
        return Optional.of(deserializeAndUpdateState(messageDataDTO.get(), buttonValue));
    }

    @VisibleForTesting
    ConfigAndState<CustomDiceConfig, StateData> deserializeAndUpdateState(@NonNull MessageDataDTO messageDataDTO, @NonNull String buttonValue) {
        Preconditions.checkArgument(CONFIG_TYPE_ID.equals(messageDataDTO.getConfigClassId()), "Unknown configClassId: %s", messageDataDTO.getConfigClassId());
        return new ConfigAndState<>(messageDataDTO.getConfigUUID(),
                Mapper.deserializeObject(messageDataDTO.getConfig(), CustomDiceConfig.class),
                new State<>(buttonValue, StateData.empty()));
    }

    @Override
    public Optional<MessageDataDTO> createMessageDataForNewMessage(@NonNull UUID configUUID,
                                                                   long guildId,
                                                                   long channelId,
                                                                   long messageId,
                                                                   @NonNull CustomDiceConfig config,
                                                                   @Nullable State<StateData> state) {
        return Optional.of(new MessageDataDTO(configUUID, guildId, channelId, messageId, getCommandId(), CONFIG_TYPE_ID,
                Mapper.serializedObject(config),
                Mapper.NO_PERSISTED_STATE, null));
    }

    @Override
    protected @NonNull String getCommandDescription() {
        return "Configure a custom set of dice buttons";
    }

    @Override
    protected @NonNull List<CommandDefinitionOption> getStartOptions() {
        return List.of(CommandDefinitionOption.builder()
                .name(BUTTONS_COMMAND_OPTIONS_ID)
                .description("Define one or more buttons separated by ';'")
                .type(CommandDefinitionOption.Type.STRING)
                .required(true)
                .build());
    }

    @Override
    protected Collection<CommandDefinitionOption> additionalCommandOptions() {
        return List.of(CommandDefinitionOption.builder()
                .name(LEGACY_START_ACTION)
                .description("Old start command")
                .type(CommandDefinitionOption.Type.SUB_COMMAND)
                .options(LEGACY_DICE_COMMAND_OPTIONS_IDS.stream()
                        .map(id -> CommandDefinitionOption.builder()
                                .name(id)
                                .description("xdy for a set of x dice with y sides, e.g. '3d6'")
                                .type(CommandDefinitionOption.Type.STRING)
                                .build())
                        .collect(Collectors.toList()))
                .option(ANSWER_TARGET_CHANNEL_COMMAND_OPTION)
                .build());
    }


    @Override
    protected @NonNull EmbedDefinition getHelpMessage() {
        return EmbedDefinition.builder()
                .description("Creates up to 25 buttons with custom dice expression e.g. '/custom_dice start buttons:3d6;10d10;3d20'. \n" + diceSystemAdapter.getHelpText(DiceParserSystem.DICE_EVALUATOR))
                .build();
    }

    @Override
    protected Set<String> getStartOptionIds() {
        return Set.of(ACTION_START, LEGACY_START_ACTION);
    }

    @Override
    protected @NonNull Optional<String> getStartOptionsValidationMessage(@NonNull CommandInteractionOption options) {
        List<String> diceExpressionWithOptionalLabel = getButtonsFromCommandOption(options).stream()
                .map(ButtonIdAndExpression::getExpression)
                .distinct()
                .collect(Collectors.toList());
        DiceParserSystem diceParserSystem = options.getName().equals(LEGACY_START_ACTION) ? DiceParserSystem.DICEROLL_PARSER : DiceParserSystem.DICE_EVALUATOR;
        return diceSystemAdapter.validateListOfExpressions(diceExpressionWithOptionalLabel, "/custom_dice help", diceParserSystem);

    }

    private List<ButtonIdAndExpression> getButtonsFromCommandOption(@NonNull CommandInteractionOption options) {
        if (LEGACY_START_ACTION.equals(options.getName())) {
            return LEGACY_DICE_COMMAND_OPTIONS_IDS.stream()
                    .flatMap(id -> options.getStringSubOptionWithName(id).stream()
                            .map(e -> new ButtonIdAndExpression(id, e)))
                    .collect(Collectors.toList());
        }
        ImmutableList.Builder<ButtonIdAndExpression> builder = ImmutableList.builder();
        String buttons = options.getStringSubOptionWithName(BUTTONS_COMMAND_OPTIONS_ID).orElseThrow();
        int idCounter = 1;
        for (String button : buttons.split(";")) {
            builder.add(new ButtonIdAndExpression(idCounter++ + "_button", button));
        }
        return builder.build();
    }

    protected @NonNull CustomDiceConfig getConfigFromStartOptions(@NonNull CommandInteractionOption options) {
        if (LEGACY_START_ACTION.equals(options.getName())) {
            BotMetrics.incrementLegacyStartCounter(getCommandId());
            return getConfigOptionStringList(getButtonsFromCommandOption(options), getAnswerTargetChannelIdFromStartCommandOption(options).orElse(null),
                    DiceParserSystem.DICEROLL_PARSER);
        }
        return getConfigOptionStringList(getButtonsFromCommandOption(options), getAnswerTargetChannelIdFromStartCommandOption(options).orElse(null), DiceParserSystem.DICE_EVALUATOR);
    }

    @VisibleForTesting
    CustomDiceConfig getConfigOptionStringList(List<ButtonIdAndExpression> startOptions, Long channelId, DiceParserSystem diceParserSystem) {
        return new CustomDiceConfig(channelId, startOptions.stream()
                .filter(be -> !be.getExpression().contains(BottomCustomIdUtils.CUSTOM_ID_DELIMITER))
                .filter(be -> !be.getExpression().contains(LABEL_DELIMITER) || be.getExpression().split(LABEL_DELIMITER).length == 2)
                .map(be -> {
                    if (be.getExpression().contains(LABEL_DELIMITER)) {
                        String[] split = be.getExpression().split(LABEL_DELIMITER);
                        return new ButtonIdLabelAndDiceExpression(be.getButtonId(), split[1].trim(), split[0].trim());
                    }
                    return new ButtonIdLabelAndDiceExpression(be.getButtonId(), be.getExpression().trim(), be.getExpression().trim());
                })
                .filter(s -> !s.getDiceExpression().isEmpty())
                .filter(s -> !s.getLabel().isEmpty())
                .filter(lv -> diceSystemAdapter.isValidExpression(lv.getDiceExpression(), diceParserSystem))
                .filter(s -> s.getDiceExpression().length() <= 2000) //limit of the discord message content
                .filter(s -> s.getLabel().length() <= 80) //https://discord.com/developers/docs/interactions/message-components#buttons
                .distinct()
                .limit(25)
                .collect(Collectors.toList()),
                diceParserSystem);
    }

    @Override
    protected @NonNull Optional<EmbedDefinition> getAnswer(CustomDiceConfig config, State<StateData> state) {
        Optional<ButtonIdLabelAndDiceExpression> selectedButton = Optional.ofNullable(state).map(State::getButtonValue)
                .flatMap(bv -> config.getButtonIdLabelAndDiceExpressions().stream()
                        .filter(bld -> bld.getButtonId().equals(bv))
                        .findFirst()
                );
        if (selectedButton.isEmpty()) {
            return Optional.empty();
        }
        //add the label only if it is different from the expression
        final String label = selectedButton.get().getDiceExpression().equals(selectedButton.get().getLabel()) ? null : selectedButton.get().getLabel();
        return Optional.of(diceSystemAdapter.answerRollWithGivenLabel(selectedButton.get().getDiceExpression(), label, false, config.getDiceParserSystem()));
    }

    @Override
    protected @NonNull Optional<MessageDefinition> createNewButtonMessageWithState(CustomDiceConfig config, State<StateData> state) {
        return Optional.of(createNewButtonMessage(config));
    }

    @Override
    public @NonNull MessageDefinition createNewButtonMessage(CustomDiceConfig config) {
        return MessageDefinition.builder()
                .content(BUTTON_MESSAGE)
                .componentRowDefinitions(createButtonLayout(config))
                .build();
    }

    private List<ComponentRowDefinition> createButtonLayout(CustomDiceConfig config) {
        List<ButtonDefinition> buttons = config.getButtonIdLabelAndDiceExpressions().stream()
                .map(d -> ButtonDefinition.builder()
                        .id(BottomCustomIdUtils.createButtonCustomId(getCommandId(), d.getButtonId()))
                        .label(d.getLabel())
                        .build())
                .collect(Collectors.toList());
        return Lists.partition(buttons, 5).stream()
                .map(bl -> ComponentRowDefinition.builder().buttonDefinitions(bl).build())
                .collect(Collectors.toList());
    }

    @Override
    protected @NonNull CustomDiceConfig getConfigFromEvent(@NonNull ButtonEventAdaptor event) {
        String[] split = event.getCustomId().split(BottomCustomIdUtils.LEGACY_CONFIG_SPLIT_DELIMITER_REGEX);
        Long answerTargetChannelId = getOptionalLongFromArray(split, 2);
        Deque<String> buttonIds = new ArrayDeque<>(IntStream.range(1, 26).mapToObj(i -> i + "_button").toList()); //legacy can have 25 buttons

        return new CustomDiceConfig(answerTargetChannelId, event.getAllButtonIds().stream()
                .map(lv -> new ButtonIdLabelAndDiceExpression(buttonIds.pop(), lv.getLabel(), BottomCustomIdUtils.getButtonValueFromLegacyCustomId(lv.getCustomId())))
                .collect(Collectors.toList()), DiceParserSystem.DICEROLL_PARSER);
    }

    @Override
    protected @NonNull State<StateData> getStateFromEvent(@NonNull ButtonEventAdaptor event) {
        final CustomDiceConfig config = getConfigFromEvent(event);
        final String buttonValue = BottomCustomIdUtils.getButtonValueFromLegacyCustomId(event.getCustomId());
        final String mappedButtonValue = config.getButtonIdLabelAndDiceExpressions().stream()
                .filter(bld -> bld.getDiceExpression().equals(buttonValue))
                .map(ButtonIdLabelAndDiceExpression::getButtonId)
                .findFirst()
                .orElseThrow();
        return new State<>(mappedButtonValue, StateData.empty());
    }

    @Override
    public String getCommandId() {
        return COMMAND_NAME;
    }

    @Value
    static class ButtonIdAndExpression {
        @NonNull
        String buttonId;
        @NonNull
        String expression;
    }

}
