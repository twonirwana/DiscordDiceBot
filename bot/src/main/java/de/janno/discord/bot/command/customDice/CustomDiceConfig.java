package de.janno.discord.bot.command.customDice;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import de.janno.discord.bot.ResultImage;
import de.janno.discord.bot.command.AnswerFormatType;
import de.janno.discord.bot.command.ButtonIdLabelAndDiceExpression;
import de.janno.discord.bot.command.Config;
import de.janno.discord.bot.dice.DiceParserSystem;
import de.janno.discord.bot.dice.image.DiceStyleAndColor;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NonNull;
import lombok.ToString;

import java.util.List;
import java.util.Locale;
import java.util.Optional;
import java.util.stream.Collectors;

@EqualsAndHashCode(callSuper = true)
@Getter
@ToString(callSuper = true)
public class CustomDiceConfig extends Config {
    @NonNull
    private final List<ButtonIdLabelAndDiceExpression> buttonIdLabelAndDiceExpressions;
    @NonNull
    private final DiceParserSystem diceParserSystem;

    @JsonCreator
    public CustomDiceConfig(@JsonProperty("answerTargetChannelId") Long answerTargetChannelId,
                            @JsonProperty("buttonIdLabelAndDiceExpressions") @NonNull List<ButtonIdLabelAndDiceExpression> buttonIdLabelAndDiceExpressions,
                            @JsonProperty("diceParserSystem") DiceParserSystem diceParserSystem,
                            @JsonProperty("answerFormatType") AnswerFormatType answerFormatType,
                            @JsonProperty("resultImage") ResultImage resultImage,
                            @JsonProperty("diceImageStyle") DiceStyleAndColor diceStyleAndColor,
                            @JsonProperty("configLocale") Locale configLocale) {
        super(answerTargetChannelId, answerFormatType, resultImage, diceStyleAndColor, configLocale);
        this.buttonIdLabelAndDiceExpressions = buttonIdLabelAndDiceExpressions;
        this.diceParserSystem = Optional.ofNullable(diceParserSystem).orElse(DiceParserSystem.DICEROLL_PARSER);
    }

    @Override
    public String toShortString() {
        String buttons = buttonIdLabelAndDiceExpressions.stream()
                .map(ButtonIdLabelAndDiceExpression::toShortString)
                .collect(Collectors.joining(", "));
        return "[%s, %s, %s, %s, %s]".formatted(buttons, getTargetChannelShortString(), diceParserSystem, getAnswerFormatType(), getDiceStyleAndColor());
    }


    @Override
    public String toCommandOptionsString() {
        String buttons = buttonIdLabelAndDiceExpressions.stream()
                .map(b -> {
                    if (b.getDiceExpression().equals(b.getLabel())) {
                        return b.getDiceExpression();
                    }
                    return "%s@%s".formatted(b.getDiceExpression(), b.getLabel());
                })
                .collect(Collectors.joining(";"));
        return "%s: %s %s".formatted(CustomDiceCommand.BUTTONS_OPTION_NAME, String.join(" ", buttons), super.toCommandOptionsString());
    }
}
