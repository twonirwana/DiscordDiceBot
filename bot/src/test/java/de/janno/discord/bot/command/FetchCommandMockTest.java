package de.janno.discord.bot.command;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import com.google.common.collect.ImmutableList;
import de.janno.discord.bot.AnswerInteractionType;
import de.janno.discord.bot.SlashEventAdaptorMock;
import de.janno.discord.bot.command.customDice.CustomDiceCommand;
import de.janno.discord.bot.command.customDice.CustomDiceConfig;
import de.janno.discord.bot.command.customParameter.CustomParameterCommand;
import de.janno.discord.bot.command.customParameter.CustomParameterConfig;
import de.janno.discord.bot.command.sumCustomSet.SumCustomSetCommand;
import de.janno.discord.bot.command.sumCustomSet.SumCustomSetConfig;
import de.janno.discord.bot.dice.CachingDiceEvaluator;
import de.janno.discord.bot.dice.image.DiceImageStyle;
import de.janno.discord.bot.dice.image.DiceStyleAndColor;
import de.janno.discord.bot.persistance.PersistenceManager;
import de.janno.discord.bot.persistance.PersistenceManagerImpl;
import de.janno.evaluator.dice.random.RandomNumberSupplier;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import java.util.List;
import java.util.Locale;
import java.util.UUID;

@ExtendWith(SnapshotExtension.class)
class FetchCommandMockTest {

    private final UUID uuid0 = UUID.fromString("00000000-0000-0000-0000-000000000000");
    private final UUID uuid1 = UUID.fromString("00000000-0000-0000-0000-000000000001");
    private Expect expect;
    private FetchCommand underTest;
    private CustomDiceCommand customDiceCommand;
    private CustomParameterCommand customParameterCommand;
    private SumCustomSetCommand sumCustomSetCommand;
    private PersistenceManager persistenceManager;
    private SlashEventAdaptorMock fetchEvent;

    @BeforeEach
    void setup() {
        persistenceManager = new PersistenceManagerImpl("jdbc:h2:mem:" + UUID.randomUUID(), null, null);
        CachingDiceEvaluator cachingDiceEvaluator = new CachingDiceEvaluator(new RandomNumberSupplier(0));
        customDiceCommand = new CustomDiceCommand(persistenceManager, cachingDiceEvaluator);
        customParameterCommand = new CustomParameterCommand(persistenceManager, cachingDiceEvaluator);
        sumCustomSetCommand = new SumCustomSetCommand(persistenceManager, cachingDiceEvaluator);
        underTest = new FetchCommand(persistenceManager, customParameterCommand, customDiceCommand, sumCustomSetCommand);
        fetchEvent = new SlashEventAdaptorMock(List.of());
    }

    @AfterEach
    void cleanup() {
        io.avaje.config.Config.setProperty("command.delayMessageDataDeletionMs", "10");
        io.avaje.config.Config.setProperty("command.fetch.delayMs", "0");
    }


    @Test
    public void getCommandDefinition() {
        expect.toMatchSnapshot(underTest.getCommandDefinition());
    }

    @Test
    public void getId() {
        expect.toMatchSnapshot(underTest.getCommandId());
    }


    @Test
    void noOldMessage() {
        underTest.handleSlashCommandEvent(fetchEvent, () -> uuid0, Locale.ENGLISH).block();

        expect.toMatchSnapshot(fetchEvent.getSortedActions());
    }

    @Test
    void noOldMessage_fr() {
        underTest.handleSlashCommandEvent(fetchEvent, () -> uuid0, Locale.FRENCH).block();

        expect.toMatchSnapshot(fetchEvent.getSortedActions());
    }

    @Test
    void noOldMessage_pt_BR() {
        underTest.handleSlashCommandEvent(fetchEvent, () -> uuid0, Locale.of("pt", "BR")).block();

        expect.toMatchSnapshot(fetchEvent.getSortedActions());
    }

    @Test
    void noOldMessage_de() {
        underTest.handleSlashCommandEvent(fetchEvent, () -> uuid0, Locale.GERMAN).block();

        expect.toMatchSnapshot(fetchEvent.getSortedActions());
    }


    @Test
    void fetchOldCustomDiceMessage() throws InterruptedException {
        io.avaje.config.Config.setProperty("command.delayMessageDataDeletionMs", "1000");
        io.avaje.config.Config.setProperty("command.fetch.delayMs", "0");

        CustomDiceConfig otherConfig = new CustomDiceConfig(null, ImmutableList.of(new ButtonIdLabelAndDiceExpression("1_button", "att", "2d20", false, false, null)), AnswerFormatType.full, AnswerInteractionType.none, null, new DiceStyleAndColor(DiceImageStyle.none, "none"), Locale.ENGLISH);
        customDiceCommand.createMessageConfig(uuid1, fetchEvent.getGuildId(), fetchEvent.getChannelId(), otherConfig)
                .ifPresent(m -> persistenceManager.saveMessageConfig(m));
        customDiceCommand.createEmptyMessageData(uuid1, fetchEvent.getGuildId(), fetchEvent.getChannelId(), -3L);

        CustomDiceConfig config = new CustomDiceConfig(null, ImmutableList.of(new ButtonIdLabelAndDiceExpression("1_button", "Dmg", "1d6", false, false, null)), AnswerFormatType.full, AnswerInteractionType.none, null, new DiceStyleAndColor(DiceImageStyle.none, "none"), Locale.ENGLISH);
        customDiceCommand.createMessageConfig(uuid0, fetchEvent.getGuildId(), fetchEvent.getChannelId(), config)
                .ifPresent(m -> persistenceManager.saveMessageConfig(m));
        customDiceCommand.createEmptyMessageData(uuid0, fetchEvent.getGuildId(), fetchEvent.getChannelId(), -2L);
        MessageDeletionHelper.deleteMessageDataWithDelay(persistenceManager, fetchEvent.getChannelId(), -2).subscribe();

        customDiceCommand.createEmptyMessageData(uuid0, fetchEvent.getGuildId(), fetchEvent.getChannelId(), -1L);
        Thread.sleep(100);

        underTest.handleSlashCommandEvent(fetchEvent, () -> uuid0, Locale.ENGLISH).block();
        expect.toMatchSnapshot(fetchEvent.getActions());
    }

    @Test
    void fetchOldCustomDiceMessageNotOldEnough() throws InterruptedException {
        io.avaje.config.Config.setProperty("command.fetch.delayMs", "6000000");
        CustomDiceConfig otherConfig = new CustomDiceConfig(null, ImmutableList.of(new ButtonIdLabelAndDiceExpression("1_button", "att", "2d20", false, false, null)), AnswerFormatType.full, AnswerInteractionType.none, null, new DiceStyleAndColor(DiceImageStyle.none, "none"), Locale.ENGLISH);
        customDiceCommand.createMessageConfig(uuid1, fetchEvent.getGuildId(), fetchEvent.getChannelId(), otherConfig)
                .ifPresent(m -> persistenceManager.saveMessageConfig(m));
        customDiceCommand.createEmptyMessageData(uuid1, fetchEvent.getGuildId(), fetchEvent.getChannelId(), -3L);

        CustomDiceConfig config = new CustomDiceConfig(null, ImmutableList.of(new ButtonIdLabelAndDiceExpression("1_button", "Dmg", "1d6", false, false, null)), AnswerFormatType.full, AnswerInteractionType.none, null, new DiceStyleAndColor(DiceImageStyle.none, "none"), Locale.ENGLISH);
        customDiceCommand.createMessageConfig(uuid0, fetchEvent.getGuildId(), fetchEvent.getChannelId(), config)
                .ifPresent(m -> persistenceManager.saveMessageConfig(m));
        customDiceCommand.createEmptyMessageData(uuid0, fetchEvent.getGuildId(), fetchEvent.getChannelId(), -2L);
        MessageDeletionHelper.deleteMessageDataWithDelay(persistenceManager, fetchEvent.getChannelId(), -2).subscribe();

        customDiceCommand.createEmptyMessageData(uuid0, fetchEvent.getGuildId(), fetchEvent.getChannelId(), -1L);

        Thread.sleep(100);
        underTest.handleSlashCommandEvent(fetchEvent, () -> uuid0, Locale.ENGLISH).block();

        expect.toMatchSnapshot(fetchEvent.getActions());
    }

    @Test
    void fetchOldSumCustomSetMessage() throws InterruptedException {
        io.avaje.config.Config.setProperty("command.delayMessageDataDeletionMs", "1000");
        io.avaje.config.Config.setProperty("command.fetch.delayMs", "0");

        SumCustomSetConfig otherConfig = new SumCustomSetConfig(null, ImmutableList.of(new ButtonIdLabelAndDiceExpression("1_button", "Att", "+2d20", false, false, null),
                new ButtonIdLabelAndDiceExpression("2_button", "bonus", "+5", false, false, null)), true, true, false, null, null, AnswerFormatType.full, AnswerInteractionType.none, null, new DiceStyleAndColor(DiceImageStyle.none, "none"), Locale.ENGLISH);
        sumCustomSetCommand.createMessageConfig(uuid1, fetchEvent.getGuildId(), fetchEvent.getChannelId(), otherConfig)
                .ifPresent(m -> persistenceManager.saveMessageConfig(m));
        sumCustomSetCommand.createEmptyMessageData(uuid1, fetchEvent.getGuildId(), fetchEvent.getChannelId(), -3L);

        SumCustomSetConfig config = new SumCustomSetConfig(null, ImmutableList.of(new ButtonIdLabelAndDiceExpression("1_button", "Dmg", "+1d6", false, false, null),
                new ButtonIdLabelAndDiceExpression("2_button", "bonus", "+2", false, false, null)), true, true, false, null, null, AnswerFormatType.full, AnswerInteractionType.none, null, new DiceStyleAndColor(DiceImageStyle.none, "none"), Locale.ENGLISH);
        sumCustomSetCommand.createMessageConfig(uuid0, fetchEvent.getGuildId(), fetchEvent.getChannelId(), config)
                .ifPresent(m -> persistenceManager.saveMessageConfig(m));
        customDiceCommand.createEmptyMessageData(uuid0, fetchEvent.getGuildId(), fetchEvent.getChannelId(), -2L);
        MessageDeletionHelper.deleteMessageDataWithDelay(persistenceManager, fetchEvent.getChannelId(), -2).subscribe();

        sumCustomSetCommand.createEmptyMessageData(uuid0, fetchEvent.getGuildId(), fetchEvent.getChannelId(), -1L);

        Thread.sleep(100);
        underTest.handleSlashCommandEvent(fetchEvent, () -> uuid0, Locale.ENGLISH).block();

        expect.toMatchSnapshot(fetchEvent.getActions());
    }

    @Test
    void fetchOldCustomParameterMessage() throws InterruptedException {
        io.avaje.config.Config.setProperty("command.delayMessageDataDeletionMs", "1000");
        io.avaje.config.Config.setProperty("command.fetch.delayMs", "0");

        CustomParameterConfig otherConfig = new CustomParameterConfig(null, "{numberOfDice:3<=>6}d{sides:6/8/10/12}", AnswerFormatType.full, AnswerInteractionType.none, null, new DiceStyleAndColor(DiceImageStyle.none, "none"), Locale.ENGLISH);
        customParameterCommand.createMessageConfig(uuid1, fetchEvent.getGuildId(), fetchEvent.getChannelId(), otherConfig)
                .ifPresent(m -> persistenceManager.saveMessageConfig(m));
        customParameterCommand.createEmptyMessageData(uuid1, fetchEvent.getGuildId(), fetchEvent.getChannelId(), -3L);

        CustomParameterConfig config = new CustomParameterConfig(null, "{numberOfDice:1<=>10}d{sides:1/4/6/8/10/12/20/100}", AnswerFormatType.full, AnswerInteractionType.none, null, new DiceStyleAndColor(DiceImageStyle.none, "none"), Locale.ENGLISH);
        customParameterCommand.createMessageConfig(uuid0, fetchEvent.getGuildId(), fetchEvent.getChannelId(), config)
                .ifPresent(m -> persistenceManager.saveMessageConfig(m));
        customParameterCommand.createEmptyMessageData(uuid0, fetchEvent.getGuildId(), fetchEvent.getChannelId(), -2L);
        MessageDeletionHelper.deleteMessageDataWithDelay(persistenceManager, fetchEvent.getChannelId(), -2).subscribe();

        customParameterCommand.createEmptyMessageData(uuid0, fetchEvent.getGuildId(), fetchEvent.getChannelId(), -1L);

        Thread.sleep(100);
        underTest.handleSlashCommandEvent(fetchEvent, () -> uuid0, Locale.ENGLISH).block();

        expect.toMatchSnapshot(fetchEvent.getActions());
    }

}