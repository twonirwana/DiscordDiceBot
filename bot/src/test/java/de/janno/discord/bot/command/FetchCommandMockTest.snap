de.janno.discord.bot.command.FetchCommandMockTest.fetchOldCustomDiceMessage=[
  [
    "reply: Fetching button message...",
    "sendMessage: EmbedOrMessageDefinition(title=null, descriptionOrContent=Click on a button to roll the dice, fields=[], componentRowDefinitions=[ComponentRowDefinition(componentDefinitions=[ButtonDefinition(label=att, id=custom_dice\u001E1_button\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null)])], hasImage=false, type=MESSAGE, userReference=false, sendToOtherChannelId=null)",
    "getMessagesState: [-3]",
    "deleteMessageById: -3"
  ]
]


de.janno.discord.bot.command.FetchCommandMockTest.fetchOldCustomDiceMessageNotOldEnough=[
  [
    "reply: No old button message in this channel found"
  ]
]


de.janno.discord.bot.command.FetchCommandMockTest.fetchOldCustomParameterMessage=[
  [
    "reply: Fetching button message...",
    "sendMessage: EmbedOrMessageDefinition(title=null, descriptionOrContent={numberOfDice}d{sides}\nPlease select value for **numberOfDice**, fields=[], componentRowDefinitions=[ComponentRowDefinition(componentDefinitions=[ButtonDefinition(label=3, id=custom_parameter\u001Eid1\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=4, id=custom_parameter\u001Eid2\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=5, id=custom_parameter\u001Eid3\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=6, id=custom_parameter\u001Eid4\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null)])], hasImage=false, type=MESSAGE, userReference=false, sendToOtherChannelId=null)",
    "getMessagesState: [-3]",
    "deleteMessageById: -3"
  ]
]


de.janno.discord.bot.command.FetchCommandMockTest.fetchOldSumCustomSetMessage=[
  [
    "reply: Fetching button message...",
    "sendMessage: EmbedOrMessageDefinition(title=null, descriptionOrContent=Click the buttons to add dice to the set and then on Roll, fields=[], componentRowDefinitions=[ComponentRowDefinition(componentDefinitions=[ButtonDefinition(label=Att, id=sum_custom_set\u001E1_button\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=bonus, id=sum_custom_set\u001E2_button\u001E00000000-0000-0000-0000-000000000001, style=PRIMARY, disabled=false, emoji=null), ButtonDefinition(label=Roll, id=sum_custom_set\u001Eroll\u001E00000000-0000-0000-0000-000000000001, style=SUCCESS, disabled=true, emoji=null), ButtonDefinition(label=Clear, id=sum_custom_set\u001Eclear\u001E00000000-0000-0000-0000-000000000001, style=DANGER, disabled=false, emoji=null), ButtonDefinition(label=Back, id=sum_custom_set\u001Eback\u001E00000000-0000-0000-0000-000000000001, style=SECONDARY, disabled=true, emoji=null)])], hasImage=false, type=MESSAGE, userReference=false, sendToOtherChannelId=null)",
    "getMessagesState: [-3]",
    "deleteMessageById: -3"
  ]
]


de.janno.discord.bot.command.FetchCommandMockTest.getCommandDefinition=[
  {
    "name": "fetch",
    "description": "Moves the last existing dice roll message in the channel to the front",
    "nameLocales": [ ],
    "descriptionLocales": [
      {
        "locale": "de",
        "description": "Bewege die letzte Würfel-Nachricht im Kanal nach vorne"
      },
      {
        "locale": "pt_BR",
        "description": "Move a última mensagem de lançamento de dados para a frente"
      },
      {
        "locale": "fr",
        "description": "Déplace le dernier message du lancer de dés vers l'avant"
      }
    ],
    "options": [ ],
    "integrationTypes": [
      "GUILD_INSTALL"
    ]
  }
]


de.janno.discord.bot.command.FetchCommandMockTest.getId=[
  "fetch"
]


de.janno.discord.bot.command.FetchCommandMockTest.noOldMessage=[
  [
    "reply: No old button message in this channel found"
  ]
]


de.janno.discord.bot.command.FetchCommandMockTest.noOldMessage_de=[
  [
    "reply: Es wurde keine alte Würfel-Nachricht im diesen Kanal gefunden"
  ]
]


de.janno.discord.bot.command.FetchCommandMockTest.noOldMessage_fr=[
  [
    "reply: Aucun ancien message de bouton trouvé"
  ]
]


de.janno.discord.bot.command.FetchCommandMockTest.noOldMessage_pt_BR=[
  [
    "reply: Nenhuma mensagem antiga do botão foi encontrada"
  ]
]